nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py

```diff
@@ -1071,3 +1080,54 @@ def on_train_batch_end(self, outputs, dataloader_iter: Any, batch_idx: int, unus
                     # Reset the optimizer update skipped to `None` - this is to prevent scheduler no-ops during
                     # accumulated gradient updates.
                     grad_scaler.optimizer_update_skipped = None
+
+    def _reset_activation_checkpointing_args(self):
+        # Store values to restore them later.
+        self.last_activations_checkpoint_granularity = self.cfg.activations_checkpoint_granularity
+        self.last_activations_checkpoint_method = self.cfg.activations_checkpoint_method
+        self.last_activations_checkpoint_num_layers = self.cfg.activations_checkpoint_num_layers
+        self.last_activations_checkpoint_layers_per_pipeline = self.cfg.activations_checkpoint_layers_per_pipeline
+
+        # Reset config values. Needed for calling generate.
+        self.cfg.activations_checkpoint_granularity = None
+        self.cfg.activations_checkpoint_method = None
+        self.cfg.activations_checkpoint_num_layers = None
+        self.cfg.activations_checkpoint_layers_per_pipeline = None
+
+        # Reset model parameters.
+        self.model.language_model.encoder.activations_checkpoint_granularity = None
+        self.model.language_model.encoder.activations_checkpoint_method = None
+        self.model.language_model.encoder.activations_checkpoint_num_layers = None
+        self.model.language_model.encoder.activations_checkpoint_layers_per_pipeline = None
+
+    def _restore_activation_checkpointing_args(self):
+        # Restore config values.
+        self.cfg.activations_checkpoint_granularity = self.last_checkpointing_granularity
+        self.cfg.activations_checkpoint_method = self.last_checkpointing_method
+        self.cfg.activations_checkpoint_num_layers = self.last_checkpointing_num_layers
+        self.cfg.activations_checkpoint_layers_per_pipeline = self.last_activations_checkpoint_layers_per_pipeline
+
+        # Restore model parameters.
+        self.model.language_model.encoder.activations_checkpoint_granularity = self.last_checkpointing_granularity
+        self.model.language_model.encoder.activations_checkpoint_method = self.last_checkpointing_method
+        self.model.language_model.encoder.activations_checkpoint_num_layers = self.last_checkpointing_num_layers
+        self.model.language_model.encoder.activations_checkpoint_layers_per_pipeline = (
+            self.last_activations_checkpoint_layers_per_pipeline
+        )
```

Comment:
I think for all `_restore_*` methods we need to default to `None` with `getattr` in case `_restore` is called before `_reset`

Reply:
Isn't this now becoming longer and the first approach which saved the values in the `original*` variables? I'm happy to do it if you think this is better, I just think it's becoming longer with this change too.

Reply:
I don't know if that's something we need to consider. `_reset` stores the config values in `last` and then sets them to `None`.
`_restore` sets the config values to `last`
If you call `_reset` they will be `None`, so `_restore` should default to `None`?
I don't think `_restore` should be used for the "original" config. `_restore` should be used for restoring from reset.

Maybe we can add docstrings to both making it clear that restore shouldn't be used before reset.

Reply:
Sorry I'm confused now. Your previous comment asked me to address the case where _restore is called before _reset. Should we handle that case or not? Sorry for the confusion, just trying to make this simple and easy to use.

Reply:
I think it's fine as is then. I think the only thing to add is a docstring explaining that `_restore` should not be called before `_reset`

Reply:
Sure, I can add the docstring.

Reply:
Docstring added @ericharper .